pages/[chain]/collection/[contract].tsx

```diff
@@ -155,6 +163,94 @@ const CollectionPage: NextPage<Props> = ({ id, ssr }) => {
     fallbackData: initialTokenFallbackData ? [ssr.tokens] : undefined,
   })
 
+  useTokenUpdateStream(id as string, collectionChain.id, {
+    onClose: () => setSocketState(0),
+    onOpen: () => setSocketState(1),
+    onMessage: ({
+      data: reservoirEvent,
+    }: MessageEvent<ReservoirWebsocketIncomingEvent>) => {
+      let hasChange = false
+
+      const newTokens = [...tokens]
+      const price = NORMALIZE_ROYALTIES
+        ? reservoirEvent.data?.market?.floorAskNormalized?.price?.amount
+            ?.decimal
+        : reservoirEvent.data?.market?.floorAsk?.price?.amount?.decimal
+      const tokenIndex = tokens.findIndex(
+        (token) => token.token?.tokenId === reservoirEvent?.data.token.tokenId
+      )
+      const token = tokenIndex > -1 ? tokens[tokenIndex] : null
+      if (token) {
+        //if the token has dynamic pricing we need to abort, this isn't supported in the websocket
+        if (token?.market?.floorAsk?.dynamicPricing) {
+          return
+        }
+        newTokens.splice(tokenIndex, 1)
+      }
+
+      if (!price) {
+        if (token) {
+          const endOfListingsIndex = tokens.findIndex(
+            (token) => !token.market?.floorAsk?.price?.amount?.decimal
+          )
+          const newTokenIndex = endOfListingsIndex > -1 ? endOfListingsIndex : 0
```

Comment:
If all tokens have a listing, we shouldn't add this token at the end because the next page may return more tokens with listings. In other words, just remove this token is endOfListingsIndex == -1. Might also be worth not removing tokens if we reached the end (can probably check the dynamic tokens hook for the hasMore property.